include/pybind11/pybind11.h

```diff
@@ -2612,13 +2612,95 @@ void print(Args &&...args) {
 }
 
 error_already_set::~error_already_set() {
-    if (m_type) {
+    if (!(m_type || m_value || m_trace)) {
+        return; // Avoid gil and scope overhead if there is nothing to release.
+    }
+    gil_scoped_acquire gil;
+    error_scope scope;
+    m_type.release().dec_ref();
+    m_value.release().dec_ref();
+    m_trace.release().dec_ref();
+}
+
+error_already_set::error_already_set(const error_already_set &e) noexcept
+    : std::exception{e}, m_lazy_what{e.m_lazy_what} {
+    gil_scoped_acquire gil;
+    m_type = e.m_type;
+    m_value = e.m_value;
+    m_trace = e.m_trace;
+}
+
+const char *error_already_set::what() const noexcept {
+    if (m_lazy_what.empty()) {
+        std::string failure_info;
```

Comment:
@rwgk Kinda of annoying to allocate space for this error string when it's so rare. Most of this could be avoided by just factoring out the custon std::terminate logic and handling of failure info into a private static method.